### PR TITLE
[wasmtime-api] Record original Trap from API callback.

### LINKED ISSUE #645: Trampolines for Callable are not registering user trap
### Repro steps

1. `cargo new repro && cd repro`
2. Add a dependency for the `wasmtime` and `wat` crates to `Cargo.toml`:

    ```toml
    wasmtime = "0.8.0"
    wat = "1.0"
    ```
    
3. Replace the contents of `src/main.rs` with the following:

    ```rust
    use std::rc::Rc;
    use wasmtime::*;
    use wat::parse_str;

    struct HelloCallback;

    impl Callable for HelloCallback {
        fn call(&self, _params: &[Val], _results: &mut [Val]) -> Result<(), HostRef<Trap>> {
            Err(HostRef::new(Trap::new("repro".into())))
        }
    }

    fn main() -> Result<(), String> {
        let engine = HostRef::new(Engine::default());
        let store = HostRef::new(Store::new(&engine));
        let binary = parse_str(
            r#"
                (module
                (func $hello (import "" "hello"))
                (func (export "run") (call $hello))
                )
            "#,
        )
        .map_err(|e| format!("failed to parse WebAssembly text source: {}", e))?;

        let module = HostRef::new(
            Module::new(&store, &binary).map_err(|e| format!("failed to compile module: {}", e))?,
        );
        let hello_type = FuncType::new(Box::new([]), Box::new([]));
        let hello_func = HostRef::new(Func::new(&store, hello_type, Rc::new(HelloCallback)));

        let imports = vec![hello_func.into()];
        let instance = Instance::new(&store, &module, imports.as_slice())
            .map_err(|e| format!("failed to instantiate module: {}", e))?;
        let run_func = instance.exports()[0]
            .func()
            .expect("expected function export");

        run_func
            .borrow()
            .call(&[])
            .map_err(|e| format!("error calling function: {:?}", e))?;

        println!("Done.");
        Ok(())
    }
    ```

4. `cargo run`

### Expected Result

The trap message should be from the returned trap:

```
Error: "error calling function: Ref(Trap { message: \"Wasm trap: repro" })"
```

### Actual Result

The trap message is from a stack overflow trap:

```
Error: "error calling function: Ref(Trap { message: \"wasm trap: call stack exhausted, source location: @-\" })"
```

### Details

I believe this is because `make_trampoline` is not registering the trap from the [conditional trap instruction](https://github.com/bytecodealliance/wasmtime/blob/master/crates/api/src/trampoline/func.rs#L150) that is conditional on the return value from `stub_fn`.

Because the trap is unregistered, a [default stack overflow trap](https://github.com/bytecodealliance/wasmtime/blob/39e57e3e9ac9c15bef45eb77a2544a7c0b76501a/crates/runtime/src/traphandlers.rs#L46-L51) is used instead of the trap code that was used for the instruction, which would be `user trap 0` in this case.

To fix this we should register the trap when compiling the trampoline.  Then the trap handler should check to see if there is a trap in the instance and use that message, otherwise falling back to the generic `user trap 0` message.

### PR DESCRIPTION
**Pull request by yurydelendik**
Fixes #645

### CODE REVIEW

**Comment by kubkon** [comment]
@yuridelendik rebasing to latest master should fix the CI for you :-) for ref, CI was fixed in #656.

File: crates/api/src/trampoline/func.rs [hunk 1]
```diff
@@ -58,12 +58,8 @@ unsafe extern "C" fn stub_fn(vmctx: *mut VMContext, call_id: u32, values_vec: *m
             0
         }
         Err(trap) => {
-            // TODO read custom exception
-            InstanceHandle::from_vmctx(vmctx)
-                .host_state()
-                .downcast_mut::<TrampolineState>()
-                .expect("state")
-                .trap = Some(trap);
+            // TODO did we record it already?
```
Comment (peterhuene) [inline crates/api/src/trampoline/func.rs]: Is this comment necessary given we assert in `record_api_trap`?

**Comment by peterhuene** [comment]
Also, this approach of returning the exact trap returned from the `Callable` is great.

In the future I think this would enable custom stack frames when we can set the frames of a trap returned from the host.

File: crates/api/src/runtime.rs [hunk 2]
```diff
@@ -112,6 +115,7 @@ pub struct Store {
     context: Context,
     global_exports: Rc<RefCell<HashMap<String, Option<wasmtime_runtime::Export>>>>,
     signature_cache: HashMap<wasmtime_runtime::VMSharedSignatureIndex, ir::Signature>,
+    trap_registration_guards: Vec<TrapRegistrationGuard>,
 }
```
Comment (peterhuene) [inline crates/api/src/runtime.rs]: Perhaps the guard should be with the `TrampolineState` along with the code memory?

Comment (yurydelendik) [reply crates/api/src/runtime.rs]: duh, attached to wrong CodeMemory, thanks

File: crates/api/src/trampoline/trap.rs [hunk 3]
```diff
@@ -0,0 +1,53 @@
+use std::cell::Cell;
+
+use super::binemit;
+use super::ir::{SourceLoc, TrapCode};
+use crate::r#ref::HostRef;
+use crate::Trap;
+use wasmtime_environ::TrapInformation;
+
+pub const API_TRAP_CODE: TrapCode = TrapCode::User(13);
```
Comment (peterhuene) [inline crates/api/src/trampoline/trap.rs]: Is there any significance to `13` here?

**Pull request merged**